From af43ed872bd64b623ea1d5b83926c4d06e8fcd7d Mon Sep 17 00:00:00 2001
From: Mike Perry <mikeperry-git@torproject.org>
Date: Fri, 7 Sep 2012 16:18:26 -0700
Subject: [PATCH 14/19] Provide an observer event to close persistent
 connections

We need to prevent linkability across "New Identity", which includes closing
keep-alive connections.
---
 netwerk/protocol/http/nsHttpHandler.cpp |    7 +++++++
 1 files changed, 7 insertions(+), 0 deletions(-)

diff --git a/netwerk/protocol/http/nsHttpHandler.cpp b/netwerk/protocol/http/nsHttpHandler.cpp
index 2f71837..b066140 100644
--- a/netwerk/protocol/http/nsHttpHandler.cpp
+++ b/netwerk/protocol/http/nsHttpHandler.cpp
@@ -309,6 +309,7 @@ nsHttpHandler::Init()
         mObserverService->AddObserver(this, "net:clear-active-logins", true);
         mObserverService->AddObserver(this, NS_PRIVATE_BROWSING_SWITCH_TOPIC, true);
         mObserverService->AddObserver(this, "net:prune-dead-connections", true);
+        mObserverService->AddObserver(this, "net:prune-all-connections", true);
         mObserverService->AddObserver(this, "net:failed-to-process-uri-content", true);
     }
  
@@ -1651,6 +1652,12 @@ nsHttpHandler::Observe(nsISupports *subject,
         if (uri && mConnMgr)
             mConnMgr->ReportFailedToProcess(uri);
     }
+    else if (strcmp(topic, "net:prune-all-connections") == 0) {
+        if (mConnMgr) {
+           mConnMgr->ClosePersistentConnections();
+           mConnMgr->PruneDeadConnections();
+        }
+    }
   
     return NS_OK;
 }
-- 
1.7.5.4

